### PR TITLE
[fuchsia][scenic] Move pointer injector to embedder.

### PR DESCRIPTION
**Pull request by shivesh-ganju**
This PR moves the pointer injector library from fuchsia views library
present in fuchsia.git to the flutter embedder.

Test: ffx test run "fuchsia-pkg://fuchsia.com/flutter_runner_tests#meta/flutter_runner_tests.cm"

### CODE REVIEW

File: shell/platform/fuchsia/flutter/engine.cc [hunk 1]
```diff
@@ -401,6 +412,7 @@ void Engine::Initialize(
            view_ref_focused = std::move(view_ref_focused),
            touch_source = std::move(touch_source),
            mouse_source = std::move(mouse_source),
+           registry = std::move(registry),
```
Comment (arbreng) [inline shell/platform/fuchsia/flutter/engine.cc:415]: nit: for readability, pointerinjector_registry here and elsewhere

File: shell/platform/fuchsia/flutter/meta/flutter_aot_product_runner.cmx [hunk 2]
```diff
@@ -26,6 +26,7 @@
             "fuchsia.ui.input.ImeService",
             "fuchsia.ui.input3.Keyboard",
             "fuchsia.ui.scenic.Scenic",
+            "fuchsia.ui.pointerinjector.Registry",
```
Comment (arbreng) [inline shell/platform/fuchsia/flutter/meta/flutter_aot_product_runner.cmx:29]: We will need to change the product session to route this service to flutter_runner in addition to the flutter app, until then connections will fail.  I would go ahead and land that before landing this, I can show you how to do it offline

File: shell/platform/fuchsia/flutter/pointer_injector_delegate.cc [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+      return false;
+    }
+
+    auto trace_flow_id = args.FindMember("traceFlowId");
+    if (!trace_flow_id->value.IsInt()) {
+      FML_LOG(ERROR) << "Argument 'traceFlowId' is not a int";
+      return false;
+    }
+
+    auto view_ref = args.FindMember("viewRef");
```
Comment (arbreng) [inline shell/platform/fuchsia/flutter/pointer_injector_delegate.cc:95]: Instead of providing ViewRef here (because its impossible under Flatland), we must harvest it in other ways.

For gfx, provide the ViewRef as an (optional) parameter to View.create
For flatland, wait on the the ChildViewWatcher.GetViewRef return value;  drop pointer events received before that point
you can store them in a std::unordered_map<uint64_t view_id, ViewRef>

Then, remove this required parameter.

File: shell/platform/fuchsia/flutter/pointer_injector_delegate.cc [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+
+bool PointerInjectorDelegate::HandlePlatformMessage(
+    rapidjson::Value request,
+    fml::RefPtr<flutter::PlatformMessageResponse> response) {
+  if (!registry_->is_bound()) {
+    FML_LOG(WARNING)
+        << "Lost connection to fuchsia.ui.pointerinjector.Registry";
+    return false;
+  }
+
```
Comment (arbreng) [inline shell/platform/fuchsia/flutter/pointer_injector_delegate.cc:43]: We need to wait here for the ViewRef to be valid, and drop any events if it is not instead of buffering them (per Jaeheon's advice)

Also, if a View was previously valid but has become invalid (because the view was destroyed or its process terminated), we need to track that in here and drop the event in that case (instead of buffering it forever below).  As a consequence, the dart code in fuchsia_views must never re-use a viewId for new views

File: shell/platform/fuchsia/flutter/pointer_injector_delegate.cc [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+      return false;
+    }
+
+    PointerInjectorRequest request = {
+        .x = pointer_x->value.GetFloat(),
+        .y = pointer_y->value.GetFloat(),
+        .pointer_id = pointer_id->value.GetUint(),
+        .phase = static_cast<fup_EventPhase>(phase->value.GetInt()),
+        .trace_flow_id = trace_flow_id->value.GetUint64(),
+        .view_ref = std::move(ref),
```
Comment (arbreng) [inline shell/platform/fuchsia/flutter/pointer_injector_delegate.cc:137]: This should come from the view_id->view_ref map

File: shell/platform/fuchsia/flutter/pointer_injector_delegate.cc [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+
+void PointerInjectorDelegate::OnCreateView(uint64_t view_id) {
+  FML_CHECK(valid_views_.count(view_id) == 0);
+
+  auto [_, success] =
+      valid_views_.try_emplace(view_id, registry_, host_view_ref_);
+
+  FML_CHECK(success);
+
+  // Dispatch any pending pointer events buffered till now for the view.
```
Comment (arbreng) [inline shell/platform/fuchsia/flutter/pointer_injector_delegate.cc:164]: there is no need; Jaeheon advised that we drop any events received before:
1. View is created
2. ViewRef is available
3. (for gfx) ViewStateChanged for that View becomes true for the first time

File: shell/platform/fuchsia/flutter/engine.cc (hunk 1 above)
Comment (shivesh-ganju) [reply shell/platform/fuchsia/flutter/engine.cc]: Done

File: shell/platform/fuchsia/flutter/meta/flutter_aot_product_runner.cmx (hunk 2 above)
Comment (shivesh-ganju) [reply shell/platform/fuchsia/flutter/meta/flutter_aot_product_runner.cmx]: Removed the CML/CMX changes for now. Will add them back in a follow up CL when the changes mentioned in the comment have been made

File: shell/platform/fuchsia/flutter/pointer_injector_delegate.cc (hunk 3 above)
Comment (shivesh-ganju) [reply shell/platform/fuchsia/flutter/pointer_injector_delegate.cc]: Done

File: shell/platform/fuchsia/flutter/pointer_injector_delegate.cc (hunk 5 above)
Comment (shivesh-ganju) [reply shell/platform/fuchsia/flutter/pointer_injector_delegate.cc]: No need. We can use OnChildViewViewRef for this

File: shell/platform/fuchsia/flutter/pointer_injector_delegate.cc (hunk 6 above)
Comment (shivesh-ganju) [reply shell/platform/fuchsia/flutter/pointer_injector_delegate.cc]: Done

File: shell/platform/fuchsia/flutter/pointer_injector_delegate.cc (hunk 4 above)
Comment (shivesh-ganju) [reply shell/platform/fuchsia/flutter/pointer_injector_delegate.cc:43]: Since we no longer require to buffer any pointer events n the case the view is not created, this problem is resolved. We will only dispatch the pointer events if the view is present in |valid_views_|. When the view is destroyed/disconnected its entry is removed from |valid_views_| and therefore all the pointer events will be dropped.

File: shell/platform/fuchsia/flutter/pointer_injector_delegate_unittest.cc [hunk 7]
```diff
@@ -0,0 +1,685 @@
...
+    }
+  }
+
+  std::unique_ptr<PointerInjectorDelegate> pointer_injector_delegate_;
+  std::unique_ptr<MockInjectorRegistry> registry_;
+  fuv_ViewRef host_view_ref_;
+  bool is_flatland_ = false;
+
+ private:
+  async::Loop loop_;
```
Comment (arbreng) [inline shell/platform/fuchsia/flutter/pointer_injector_delegate_unittest.cc:248]: Prefer to use async::TestLoop here, along with a subloop, via #include <lib/async-testing/test_loop.h> from the SDK

It gives you RunLoopUntil & friends for free, and FIDL works, all in a very deterministic way.  It even allows you to control the loop time, to test exotic edge cases!  See flatland_connection_unittest.cc for example uasge

**Review by arbreng — APPROVED** [review]

File: shell/platform/fuchsia/flutter/pointer_injector_delegate.cc [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+        << "Lost connection to fuchsia.ui.pointerinjector.Registry";
+    return false;
+  }
+
+  auto method = request.FindMember("method");
+  if (method == request.MemberEnd() || !method->value.IsString()) {
+    return false;
+  }
+
+  if (method->value == "View.Pointerinjector.inject") {
```
[the remark this answers is not in the document]
Comment (shivesh-ganju) [reply shell/platform/fuchsia/flutter/pointer_injector_delegate.cc:49]: Done

File: shell/platform/fuchsia/flutter/platform_view.cc [hunk 9]
```diff
@@ -845,6 +855,9 @@ bool PlatformView::HandleFlutterPlatformViewsChannelPlatformMessage(
     }
   } else if (method.rfind("View.focus", 0) == 0) {
     return focus_delegate_->HandlePlatformMessage(root, message->response());
+  } else if (method.rfind("View.Pointerinjector", 0) == 0) {
```
[the remark this answers is not in the document]
Comment (shivesh-ganju) [reply shell/platform/fuchsia/flutter/platform_view.cc:858]: Done

File: shell/platform/fuchsia/flutter/pointer_injector_delegate_unittest.cc (hunk 7 above)
Comment (shivesh-ganju) [reply shell/platform/fuchsia/flutter/pointer_injector_delegate_unittest.cc:250]: Filed a bug for the same. fxb/104285
